File: src/main/java/commands/core/CommandHandler.java
package commands.core;

import commands.DiscordCommands;
import util.commands.AnnotationExtractor;
import util.commands.DiscordCommand;
import util.MessageSending;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sx.blah.discord.handle.obj.IMessage;

import java.util.*;
import java.util.regex.Pattern;

public class CommandHandler implements DiscordCommands{
	private Map<String, Command> commands = new HashMap<>();
	//TODO make this a config option
	private String PREFIX = "!";
	private Logger logger = LoggerFactory.getLogger(CommandHandler.class);
	private StringBuilder cmdPattern = new StringBuilder();
	public CommandHandler() {
	}

	public void registerCommand(Command cmd) {
		commands.put(cmd.getName(), cmd);
		if (cmdPattern.length() > 1) {
			cmdPattern.append('|');
		}
		cmdPattern.append(Pattern.quote(cmd.getName()));
	}

	public void registerCommand(String commandName, Command cmd) {
		commands.put(commandName, cmd);
		if (cmdPattern.length() > 1) {
			cmdPattern.append('|');
		}
		cmdPattern.append(Pattern.quote(commandName));
	}

	public void unregisterCommand(String name) {
		if (!commands.containsKey(name)) throw new IllegalArgumentException(name + " is not a valid command");
		commands.remove(name);
		name = Pattern.quote(name);
		//This craziness needs to be in place in order to allow commands to contain all characters including ones that are used by regex
		cmdPattern.replace(cmdPattern.indexOf(name), cmdPattern.indexOf(name) + name.length(), "");

		if (cmdPattern.charAt(0) == '|') {
			cmdPattern.deleteCharAt(0);
		} else if (cmdPattern.charAt(cmdPattern.length() - 1) == '|') {
			cmdPattern.deleteCharAt(cmdPattern.length() - 1);
		} else {
			int check = cmdPattern.indexOf("||");
			if (check > 0) {
				cmdPattern.replace(check, check + 2, "|");
			}
		}

	}

	public void runCommand(IMessage message, String[] digestedString) {
		if (!message.getChannel().isPrivate() && isCommand(message.getContent())) {
			if (commands.containsKey(digestedString[0])) {

				logger.debug("Executing command '{}'", digestedString[0]);
				//System.out.printf("[CommandHandler] Executing command '%s'\r\n", digestedString[0]);
				commands.get(digestedString[0]).execute(message, digestedString);
			} /*else {
				message.getChannel().sendMessage("Command `" + PREFIX + digestedString[0] + "` not found.");
			}*/

		}
	}

	@DiscordCommand("list")
	public void list(IMessage msg, String...args) {
		Iterator<String> iterator = commands.keySet().iterator();
		String commands = "";
		StringBuilder builder = new StringBuilder();
		while (iterator.hasNext()) {
			builder.append('!').append(iterator.next());
			if (iterator.hasNext()) {
				builder.append(", ");
			} else {
				commands = builder.toString();
			}
		}
		MessageSending.sendMessage(msg.getChannel(), "Commands are: " + commands);

	}

	public Command getCommandByName(String name) {
		return commands.get(name);
	}

	public Collection<Command> getAllCommands() {
		return commands.values();
	}

	public boolean isCommand(String str) {
		return str.matches(" *" + Pattern.quote(PREFIX) + '(' +cmdPattern.toString() + ')' + ".*");
	}
	public String[] getArguments(String input) {
		return input.substring(input.indexOf(PREFIX) + 1).split(" +");
	}

	@Override
	public Collection<Command> getDiscordCommands() {
		return AnnotationExtractor.getCommandList(this);
	}
}

File: src/main/kotlin/usbbot/commands/CommandHandler.kt
package usbbot.commands

import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.newFixedThreadPoolContext
import kotlinx.coroutines.experimental.newSingleThreadContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import usbbot.commands.core.Command
import usbbot.commands.security.PermissionManager
import usbbot.modules.SimpleTextResponses
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.Permissions
import usbbot.config.*
import usbbot.util.MessageSending
import kotlin.system.measureTimeMillis

class CommandHandler {
    companion object {
        val logger : Logger = LoggerFactory.getLogger(CommandHandler::class.java)
        val stc = newFixedThreadPoolContext(4,"commandExecutor")
    }
    private var cmdMap = HashMap<String, Command>()
    init {
        cmdMap.put("list", object: Command() {
            override fun execute(msg: IMessage, args: Array<String>) {
                val iterator = cmdMap.keys.iterator()
                val builder = StringBuilder()
                while (iterator.hasNext()) {
                    builder.append('`').append(iterator.next()).append("`").append(", ")
                }

                var iterator2 : Iterator<DBTextCommand>? = null
                val time = measureTimeMillis {
                    iterator2 = getDBTextCommandsForGuild(msg.guild.longID).iterator()
                }
                logger.trace("It took me {}ms to look up how many text commands there are.", time)

                while (iterator2!!.hasNext()) {
                    builder.append('`').append(iterator2!!.next().name).append("`")
                    if (iterator2!!.hasNext()) {
                        builder.append(", ")
                    }
                }
                MessageSending.sendMessage(msg.channel, "Commands are: " + builder.toString())
            }
        })
    }
    fun createMissingPermissions(guildID: Long) {
        var guildCommands = getCommandsForGuild(guildID)
        cmdMap.keys.filter { cmdName -> guildCommands.find { it.name == cmdName} == null }.forEach {
            createDBCommand(guildID, it, "whitelist", "whitelist")
        }
    }
    fun registerCommand(cmd: Command) = cmdMap.put(cmd.name, cmd)
    fun registerCommands(cmds: DiscordCommands) = cmds.discordCommands.forEach({registerCommand(it)})
    fun discordCommandExists(name: String, guildID: Long) : Boolean {
        if (cmdMap.containsKey(name)) return true
        //FIXME: This only exists so that the SimpleTextResponses Module knows that a command with that name is already registered, so remove this
        if (getCommandForGuild(guildID, name) != null) return true
        return false
    }
    fun getArguments(input: String, prefix: String): Array<String> {
        return input.substring(input.indexOf(prefix) + prefix.length).split(" +".toRegex()).dropWhile({ it.isEmpty() }).toTypedArray()
    }
    fun onMessageRecivedEvent(event: MessageReceivedEvent) {
        //Check that the message was not send in a private channel, if it was just ignore it.
        //repeat(10) {
            val timeForCoroutineStart = measureTimeMillis {
                launch(stc) {
                    val timeCoroutineStart = System.currentTimeMillis()
                        if (!(event.author.isBot || event.channel.isPrivate)) {
                            //TODO: Check if the message is on the word/regex blacklist, remove it if it is (blacklist may not apply to all users)
                            //Check if the message starts with the server command prefix, if not ignore it
                            val guild = getGuildById(event.guild.longID) ?:
                                    throw IllegalStateException("A Command was tried to be executed on a Guild that has no DB entry")

                            if (event.message.content.startsWith(guild.prefix)) {
                                val args = getArguments(event.message.content, guild.prefix)
                                //check if the message contains a valid command for that guild and check permissions
                                //I need to check if the command exists before testing for permissions, otherwise a permission entry will be created
                                logger.trace("IT took me {}ms to get before StupidWrapper", System.currentTimeMillis() - timeCoroutineStart)
                                val stupidWrapper = StupidWrapper()

                                logger.trace("IT took me {}ms to get after StupidWrapper", System.currentTimeMillis() - timeCoroutineStart)
                                if(stupidWrapper.isCommand(cmdMap, event.guild.longID, args[0])) {

                                    logger.trace("IT took me {}ms to check if it is a command", System.currentTimeMillis() - timeCoroutineStart)
                                    val isAdministrator = event.author.getPermissionsForGuild(event.guild).contains(Permissions.ADMINISTRATOR)

                                    if(isAdministrator || PermissionManager.hasPermission(
                                            event.guild.longID,
                                            event.author.longID,
                                            event.message.guild.getRolesForUser(event.author).map { it.longID },
                                            args[0])) {

                                        logger.trace("IT took me {}ms to check permissions (successful)", System.currentTimeMillis() - timeCoroutineStart)
                                        if (stupidWrapper.stc == null) {
                                            cmdMap[args[0]]?.execute(event.message, args)
                                        } else {
                                            SimpleTextResponses.answer(event.message, stupidWrapper.stc)
                                        }
                                    } else {
                                        logger.trace("IT took me {}ms to check permissions (unsuccessful)", System.currentTimeMillis() - timeCoroutineStart)
                                        MessageSending.sendMessage(event.channel, "You don't have permissions required to use this command!")
                                    }
                                    logger.trace("IT took me {}ms to run the command", System.currentTimeMillis() - timeCoroutineStart)
                                }
                            }
                        }
                    logger.trace("It took me {} ms to execute the Coroutine for message {}", System.currentTimeMillis() - timeCoroutineStart, event.message.content)
                }
            }
            logger.trace("It took me {} ms to start the Coroutine for message {}", timeForCoroutineStart, event.message.content)
    //    }


    }

}

private class StupidWrapper {
    var stc : DBTextCommand? = null
    fun isCommand(cmdMap: HashMap<String, Command>, guildID: Long, name: String) : Boolean {
        if (cmdMap.containsKey(name)) return true
        stc = getDBTextCommand(guildID, name)
        if (stc != null) return true
        return false
    }
}
